### PR TITLE
[Merged by Bors] - chore: deprecate `Init.Data.Nat.Lemmas`

### PR DESCRIPTION
**Pull request by Parcly-Taxel**

### CODE REVIEW

**Review by YaelDillies — COMMENTED** [review]
We definitely want to keep `Nat.twoStepInduction`

File: Mathlib/Init/Data/Nat/Lemmas.lean [hunk 1]
```diff
@@ -69,14 +61,15 @@ def subInduction {P : ℕ → ℕ → Sort u} (H1 : ∀ m, P 0 m) (H2 : ∀ n, P
   | succ n, succ m => H3 _ _ (subInduction H1 H2 H3 n m)
 
 -- Porting note: added `elab_as_elim`
-@[elab_as_elim]
+@[elab_as_elim, deprecated Nat.strongRecOn (since := "2024-07-27")]
 protected theorem strong_induction_on {p : Nat → Prop} (n : Nat)
```
Comment (j-loreaux) [inline Mathlib/Init/Data/Nat/Lemmas.lean:65]: most of this PR is just dealing with the fallout of this deprecation, but I don't see the point. I mean, I realize that `strongRecOn` is `Sort`-valued as opposed to `Prop`-valued, but why not just add a docstring to `strong_induction_on` which says: "This is for proving theorems. If you need to construct data, then use `Nat.strongRecOn`."

Then just move `Nat.strong_induction_on` to a different file.

Comment (Parcly-Taxel) [reply Mathlib/Init/Data/Nat/Lemmas.lean]: Aha. `Nat.(cases_)strong_induction_on` should still be deprecated, but to `Nat.(cases)strongInductionOn` in `Init.WF`, which I have done. Besides, why would I want to keep a theorem that is a **strictly less general** version of another one and is therefore redundant?

Comment (YaelDillies) [reply Mathlib/Init/Data/Nat/Lemmas.lean]: I am of the opinion that `Nat.strong_induction_on` should be the Prop-valued specialisation of `Nat.strongRecOn`

File: Mathlib/Algebra/ContinuedFractions/Computation/Approximations.lean [hunk 2]
```diff
@@ -205,8 +205,8 @@ denominators `Bₙ` are bounded from below by the fibonacci sequence `Nat.fib`.
 open Nat
 
 theorem fib_le_of_contsAux_b :
-    n ≤ 1 ∨ ¬(of v).TerminatedAt (n - 2) → (fib n : K) ≤ ((of v).contsAux n).b :=
-  Nat.strong_induction_on n
+    n ≤ 1 ∨ ¬(of v).TerminatedAt (n - 2) → (fib n : K) ≤ ((of v).contsAux n).b := by
+  apply Nat.strongInductionOn n
```
Comment (YaelDillies) [inline Mathlib/Algebra/ContinuedFractions/Computation/Approximations.lean:209]: Please revert

File: Archive/Imo/Imo1981Q3.lean [hunk 3]
```diff
@@ -108,7 +108,7 @@ theorem n_le_N {m n : ℕ} (h1 : NatPredicate N m n) : n ≤ N := mod_cast h1.n_
 Now we can use induction to show that solutions must be Fibonacci numbers.
 -/
 theorem imp_fib {n : ℕ} : ∀ m : ℕ, NatPredicate N m n → ∃ k : ℕ, m = fib k ∧ n = fib (k + 1) := by
-  refine Nat.strong_induction_on n ?_
+  apply Nat.strongInductionOn n
```
Comment (YaelDillies) [inline Archive/Imo/Imo1981Q3.lean:111]: Please revert

**Review by YaelDillies — COMMENTED** [review]
etc...

File: Mathlib/Init/Data/Nat/Lemmas.lean (hunk 1 above)
Comment (Parcly-Taxel) [reply Mathlib/Init/Data/Nat/Lemmas.lean]: @YaelDillies do you want me to add the Prop-valued specialisations to my core PR?

File: Mathlib/Algebra/ContinuedFractions/Computation/Approximations.lean (hunk 2 above)
Comment (Parcly-Taxel) [reply Mathlib/Algebra/ContinuedFractions/Computation/Approximations.lean:209]: If I reverted there would be an error

File: Archive/Imo/Imo1981Q3.lean (hunk 3 above)
Comment (Parcly-Taxel) [reply Archive/Imo/Imo1981Q3.lean]: See my reply above.

File: Mathlib/Init/Data/Nat/Lemmas.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Init/Data/Nat/Lemmas.lean]: I think it's fine to have `strong_induction_on` in mathlib only. If people want to upstream it, they always can

File: Mathlib/Order/Defs.lean [hunk 4]
```diff
@@ -390,6 +390,22 @@ def ltByCases (x y : α) {P : Sort*} (h₁ : x < y → P) (h₂ : x = y → P) (
...
+
+@[deprecated (since := "2024-08-23")]
+protected def ltGeByCases {a b : Nat} {C : Sort u} (h₁ : a < b → C) (h₂ : b ≤ a → C) : C :=
+  Decidable.byCases h₁ fun h => h₂ (Or.elim (Nat.lt_or_ge a b) (fun a => absurd a h) fun a => a)
+
+set_option linter.deprecated false in
+@[deprecated ltByCases (since := "2024-08-23")]
+protected def ltByCases {a b : Nat} {C : Sort u} (h₁ : a < b → C) (h₂ : a = b → C)
+    (h₃ : b < a → C) : C :=
+  Nat.ltGeByCases h₁ fun h₁ => Nat.ltGeByCases h₃ fun h => h₂ (Nat.le_antisymm h h₁)
```
Comment (jcommelin) [inline Mathlib/Order/Defs.lean:405]: Why are these deprecated? What is replacing them?

Comment (Parcly-Taxel) [reply Mathlib/Order/Defs.lean:405]: `Nat.ltGeByCases` does not have a replacement, but `Nat.ltByCases` is being replaced by `ltByCases` (without namespace).

File: Archive/Arithcc.lean
Comment (YaelDillies) [inline Archive/Arithcc.lean:1]: All the changes in this file (except the very import change are scope creep. Please revert

File: Archive/Imo/Imo1981Q3.lean
Comment (YaelDillies) [inline Archive/Imo/Imo1981Q3.lean:1]: Please revert

File: Mathlib/Algebra/ContinuedFractions/Computation/Approximations.lean
Comment (YaelDillies) [inline Mathlib/Algebra/ContinuedFractions/Computation/Approximations.lean:1]: Please revert

File: Mathlib/Algebra/ContinuedFractions/Computation/TerminatesIffRat.lean
Comment (YaelDillies) [inline Mathlib/Algebra/ContinuedFractions/Computation/TerminatesIffRat.lean:1]: Please revert

File: Mathlib/Algebra/ContinuedFractions/ConvergentsEquiv.lean
Comment (YaelDillies) [inline Mathlib/Algebra/ContinuedFractions/ConvergentsEquiv.lean:1]: Please revert

Comment (YaelDillies) [inline Mathlib/Algebra/ContinuedFractions/ConvergentsEquiv.lean:1]: Please revert

File: Mathlib/Algebra/Polynomial/Derivative.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Polynomial/Derivative.lean:1]: Please revert

File: Mathlib/Algebra/Polynomial/Lifts.lean
Comment (YaelDillies) [inline Mathlib/Algebra/Polynomial/Lifts.lean:1]: Please revert

File: Mathlib/Analysis/Analytic/Basic.lean
Comment (YaelDillies) [inline Mathlib/Analysis/Analytic/Basic.lean:1]: Please revert

File: Mathlib/Analysis/Calculus/ContDiff/Bounds.lean
Comment (YaelDillies) [inline Mathlib/Analysis/Calculus/ContDiff/Bounds.lean:1]: Please revert

File: Mathlib/Analysis/Hofer.lean
Comment (YaelDillies) [inline Mathlib/Analysis/Hofer.lean:1]: Please revert

File: Mathlib/Combinatorics/Derangements/Finite.lean
Comment (YaelDillies) [inline Mathlib/Combinatorics/Derangements/Finite.lean:1]: Please revert

File: Mathlib/Combinatorics/Enumerative/Catalan.lean
Comment (YaelDillies) [inline Mathlib/Combinatorics/Enumerative/Catalan.lean:1]: Please revert

File: Mathlib/Combinatorics/Hall/Finite.lean
Comment (YaelDillies) [inline Mathlib/Combinatorics/Hall/Finite.lean:1]: Please revert

File: Mathlib/Combinatorics/SetFamily/AhlswedeZhang.lean
Comment (YaelDillies) [inline Mathlib/Combinatorics/SetFamily/AhlswedeZhang.lean:1]: Please revert

File: Mathlib/Computability/Primrec.lean
Comment (YaelDillies) [inline Mathlib/Computability/Primrec.lean:1]: Please revert

File: Mathlib/Data/Fintype/Basic.lean
Comment (YaelDillies) [inline Mathlib/Data/Fintype/Basic.lean:1]: Please revert, except for the import change obviously

File: Mathlib/Data/Fintype/Card.lean
Comment (YaelDillies) [inline Mathlib/Data/Fintype/Card.lean:1]: Please revert

File: Mathlib/Data/Nat/Choose/Central.lean
Comment (YaelDillies) [inline Mathlib/Data/Nat/Choose/Central.lean:1]: Please revert

File: Mathlib/Data/Nat/Digits.lean
Comment (YaelDillies) [inline Mathlib/Data/Nat/Digits.lean:1]: Please revert

File: Mathlib/Data/Nat/Log.lean
Comment (YaelDillies) [inline Mathlib/Data/Nat/Log.lean:1]: Please revert, except for the import change obviously

File: Mathlib/FieldTheory/Separable.lean
Comment (YaelDillies) [inline Mathlib/FieldTheory/Separable.lean:1]: Please revert

File: Mathlib/Init/Data/Nat/Lemmas.lean [hunk 5]
```diff
@@ -68,18 +53,15 @@ def subInduction {P : ℕ → ℕ → Sort u} (H1 : ∀ m, P 0 m) (H2 : ∀ n, P
   | succ _n, 0 => H2 _
   | succ n, succ m => H3 _ _ (subInduction H1 H2 H3 n m)
 
--- Porting note: added `elab_as_elim`
-@[elab_as_elim]
+@[elab_as_elim, deprecated Nat.strongInductionOn (since := "2024-08-23")]
 protected theorem strong_induction_on {p : Nat → Prop} (n : Nat)
     (h : ∀ n, (∀ m, m < n → p m) → p n) : p n :=
```
Comment (YaelDillies) [inline Mathlib/Init/Data/Nat/Lemmas.lean:58]: Please revert and instead move `strong_induction_on` to `Data.Nat.Defs`

File: Mathlib/LinearAlgebra/Eigenspace/Triangularizable.lean
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/Eigenspace/Triangularizable.lean:1]: Please revert

File: Mathlib/LinearAlgebra/Multilinear/Basic.lean
Comment (YaelDillies) [inline Mathlib/LinearAlgebra/Multilinear/Basic.lean:1]: Please revert

File: Mathlib/NumberTheory/BernoulliPolynomials.lean
Comment (YaelDillies) [inline Mathlib/NumberTheory/BernoulliPolynomials.lean:1]: Please revert

File: Mathlib/NumberTheory/DiophantineApproximation.lean
Comment (YaelDillies) [inline Mathlib/NumberTheory/DiophantineApproximation.lean:1]: Please revert

File: Mathlib/NumberTheory/Pell.lean
Comment (YaelDillies) [inline Mathlib/NumberTheory/Pell.lean:1]: Please revert

File: Mathlib/NumberTheory/Primorial.lean
Comment (YaelDillies) [inline Mathlib/NumberTheory/Primorial.lean:1]: Please revert

File: Mathlib/Order/Fin/Basic.lean
Comment (YaelDillies) [inline Mathlib/Order/Fin/Basic.lean:1]: Please revert, except for the import change obviously

File: Mathlib/RingTheory/Adjoin/PowerBasis.lean
Comment (YaelDillies) [inline Mathlib/RingTheory/Adjoin/PowerBasis.lean:1]: Please revert

File: Mathlib/RingTheory/LittleWedderburn.lean
Comment (YaelDillies) [inline Mathlib/RingTheory/LittleWedderburn.lean:1]: Please revert

File: Mathlib/RingTheory/Polynomial/Basic.lean
Comment (YaelDillies) [inline Mathlib/RingTheory/Polynomial/Basic.lean:1]: Please revert

File: Mathlib/RingTheory/Polynomial/Cyclotomic/Basic.lean
Comment (YaelDillies) [inline Mathlib/RingTheory/Polynomial/Cyclotomic/Basic.lean:1]: Please revert

File: Mathlib/RingTheory/Polynomial/Cyclotomic/Eval.lean
Comment (YaelDillies) [inline Mathlib/RingTheory/Polynomial/Cyclotomic/Eval.lean:1]: Please revert

File: Mathlib/RingTheory/Polynomial/Eisenstein/IsIntegral.lean
Comment (YaelDillies) [inline Mathlib/RingTheory/Polynomial/Eisenstein/IsIntegral.lean:1]: Please revert

File: Mathlib/RingTheory/Polynomial/Nilpotent.lean
Comment (YaelDillies) [inline Mathlib/RingTheory/Polynomial/Nilpotent.lean:1]: Please revert

File: Mathlib/RingTheory/PowerSeries/Basic.lean
Comment (YaelDillies) [inline Mathlib/RingTheory/PowerSeries/Basic.lean:1]: Please revert

File: Mathlib/RingTheory/QuotientNilpotent.lean
Comment (YaelDillies) [inline Mathlib/RingTheory/QuotientNilpotent.lean:1]: Please revert

File: Mathlib/RingTheory/WittVector/Defs.lean
Comment (YaelDillies) [inline Mathlib/RingTheory/WittVector/Defs.lean:1]: Please revert

File: Mathlib/RingTheory/WittVector/Frobenius.lean
Comment (YaelDillies) [inline Mathlib/RingTheory/WittVector/Frobenius.lean:1]: Please revert

File: Mathlib/RingTheory/WittVector/StructurePolynomial.lean
Comment (YaelDillies) [inline Mathlib/RingTheory/WittVector/StructurePolynomial.lean:1]: Please revert

File: Mathlib/SetTheory/Game/Nim.lean
Comment (YaelDillies) [inline Mathlib/SetTheory/Game/Nim.lean:1]: Please revert

File: Mathlib/SetTheory/Surreal/Dyadic.lean
Comment (YaelDillies) [inline Mathlib/SetTheory/Surreal/Dyadic.lean:1]: Please revert

File: Mathlib/Topology/Metrizable/Uniformity.lean
Comment (YaelDillies) [inline Mathlib/Topology/Metrizable/Uniformity.lean:1]: Please revert

**Review by YaelDillies — COMMENTED** [review]
Since you seem to not differentiate what is scope-creep from what's not, I have commented a full list of what I see as scope-creep in this PR

**Comment by YaelDillies** [comment]
Thanks!

maintainer merge

**Comment by semorrison** [comment]
bors merge
